File: 653.py
from tkinter.tix import Tree
from typing import Optional


# Definition for a binary tree node.
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        # 这个写法太妙了，很好地利用了Python的生成器，不用自己构造栈模拟左右指针的移动
        def inorder(root: TreeNode):
            if not root:
                return
            yield from inorder(root.left)
            yield root.val
            yield from inorder(root.right)

        def inorder_reverse(root: TreeNode):
            if not root:
                return
            yield from inorder_reverse(root.right)
            yield root.val
            yield from inorder_reverse(root.left)

        left_gen, right_gen = inorder(root), inorder_reverse(root)
        left, right = next(left_gen), next(right_gen)
        while left < right:
            print(left, right)
            if (tmp := left + right) == k:
                return True
            elif tmp > k:
                right = next(right_gen)
            else:
                left = next(left_gen)
        return False


root = TreeNode(2)
root.left = TreeNode(1)
root.right = TreeNode(3)

print(Solution().findTarget(root, 3))

File: 653.go
package leetcode

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findTarget(root *TreeNode, k int) bool {
	left, right := root, root
	// left指向最小元素, right指向最大元素
	left_stack, right_stack := []*TreeNode{left}, []*TreeNode{right}

	for left.Left != nil {
		left = left.Left
		left_stack = append(left_stack, left)
	}

	for right.Right != nil {
		right = right.Right
		right_stack = append(right_stack, right)
	}

	for left != right {
		if tmp := left.Val + right.Val; tmp == k {
			return true
		} else if tmp < k {
			left_stack = left_stack[:len(left_stack)-1]
			node := left.Right
			for node != nil {
				left_stack = append(left_stack, node)
				node = node.Left
			}
			left = left_stack[len(left_stack)-1]
		} else {
			right_stack = right_stack[:len(right_stack)-1]
			node := right.Left
			for node != nil {
				right_stack = append(right_stack, node)
				node = node.Right
			}
			right = right_stack[len(right_stack)-1]
		}
	}
	return false
}
